File: examples/toolqry.c
/*
 * Copyright (c) 2025      Nanook Consulting  All rights reserved.
 * $COPYRIGHT$
 *
 * Additional copyrights may follow
 *
 * $HEADER$
 *
 */

#define _GNU_SOURCE
#include <stdio.h>
#include <libgen.h>

#include "examples.h"
#include <pmix_tool.h>

static void querycbfunc(pmix_status_t status, pmix_info_t *info, size_t ninfo, void *cbdata,
                        pmix_release_cbfunc_t release_fn, void *release_cbdata)
{
    myquery_data_t *mq = (myquery_data_t *) cbdata;
    size_t n;

    mq->lock.status = status;

    /* save the returned info - it will be
     * released in the release_fn */
    if (0 < ninfo) {
        PMIX_INFO_CREATE(mq->info, ninfo);
        mq->ninfo = ninfo;
        for (n = 0; n < ninfo; n++) {
            PMIX_INFO_XFER(&mq->info[n], &info[n]);
        }
    }

    /* let the library release the data */
    if (NULL != release_fn) {
        release_fn(release_cbdata);
    }

    /* release the block */
    DEBUG_WAKEUP_THREAD(&mq->lock);
}

int main(int argc, char **argv)
{
	pmix_status_t rc;
	char *kptr;
	pmix_rank_t rank = 0;
	pmix_info_t info[3], *iptr, *ip;
	pmix_proc_t myproc, proc;
	char hostname[1024];
	pmix_value_t *val = NULL;
	char *toolname;
    pmix_query_t *query;
    myquery_data_t mydata;
    size_t nq, ninfo, sz, m;
    pmix_data_array_t *darray, *dptr;
    int i;
    bool system_server = false;

    // check for directives
    for (i=1; i < argc; i++) {
        if (0 == strcmp(argv[i], "--system-server")) {
            system_server = true;
        }
    }

    /* assign our own name */
    ninfo = 2;
    gethostname(hostname, sizeof(hostname));
    toolname = basename(argv[0]);
    rc = asprintf(&kptr, "%s.%s.%lu", toolname, hostname, (unsigned long)getpid());
    if (0 > rc) {
        fprintf(stderr, "Failed to create name\n");
        exit(1);
    }
    PMIX_INFO_LOAD(&info[0], PMIX_TOOL_NSPACE, kptr, PMIX_STRING);
    free(kptr);
    PMIX_INFO_LOAD(&info[1], PMIX_TOOL_RANK, &rank, PMIX_PROC_RANK);
    // if they want us to use a system server, then say so
    if (system_server) {
        PMIX_INFO_LOAD(&info[2], PMIX_CONNECT_TO_SYSTEM, NULL, PMIX_BOOL);
        ninfo = 3;
    }

    /* init as a tool */
    rc = PMIx_tool_init(&myproc, info, ninfo);
    if (PMIX_SUCCESS != rc) {
        fprintf(stderr, "PMIx_tool_init failed: %s\n", PMIx_Error_string(rc));
        exit(rc);
    }
    PMIX_PROC_CONSTRUCT(&proc);

    /* query the list of active nspaces */
    nq = 1;
    PMIX_QUERY_CREATE(query, nq);
    PMIX_ARGV_APPEND(rc, query[0].keys, PMIX_QUERY_NAMESPACE_INFO);
    DEBUG_CONSTRUCT_MYQUERY(&mydata);
    if (PMIX_SUCCESS != (rc = PMIx_Query_info_nb(query, nq, querycbfunc, (void *) &mydata))) {
        fprintf(stderr, "Client ns %s rank %d: PMIx_Query_info failed: %d\n", myproc.nspace,
                myproc.rank, rc);
        goto done;
    }
    DEBUG_WAIT_THREAD(&mydata.lock);
    rc = mydata.lock.status;
    /* find the response */
    if (PMIX_SUCCESS == rc) {
        /* should be in the first key */
        if (PMIX_CHECK_KEY(&mydata.info[0], PMIX_QUERY_NAMESPACE_INFO)) {
            darray = mydata.info[0].value.data.darray;
            if (NULL == darray || 0 == darray->size || NULL == darray->array) {
                fprintf(stderr, "No namespaces active\n");
                rc = PMIX_ERROR;
                goto done;
            } else {
                ip = (pmix_info_t *) darray->array;
                sz = darray->size;
                if (NULL == ip || 0 == sz) {
                    fprintf(stderr, "Error\n");
                    rc = PMIX_ERROR;
                    goto done;
                } else {
                    // find our namespace
                    for (m=0; m < sz; m++) {
                        dptr = ip[m].value.data.darray;
                        if (NULL == dptr || NULL == dptr->array || 0 == dptr->size) {
                            fprintf(stderr, "Error\n");
                            rc = PMIX_ERROR;
                            goto done;
                        }
                        iptr = (pmix_info_t *) dptr->array;
                        // looking for nspaces with an "@N" at the end
                        kptr = strrchr(iptr[0].value.data.string, '@');
                        if (NULL == kptr) {
                            continue;
                        }
                        ++kptr;
                        // ignore the "0" namespace as that is the DVM
                        i = strtol(kptr, NULL, 10);
                        if (0 == i) {
                            continue;
                        }
                        // take this one
                        PMIX_LOAD_PROCID(&proc, iptr[0].value.data.string, 0);
                        fprintf(stderr, "Received nspace %s\n", proc.nspace);
                        break;
                    }
                }
            }
        } else {
            fprintf(stderr, "Query returned wrong info key at first posn: %s\n",
                    mydata.info[0].key);
            rc = PMIX_ERROR;
            goto done;
        }
    } else {
        fprintf(stderr, "Query returned error: %s\n", PMIx_Error_string(mydata.lock.status));
        goto done;
    }
    DEBUG_DESTRUCT_MYQUERY(&mydata);

    rc = PMIx_Get(&proc, PMIX_APPNUM, NULL, 0, &val);
    fprintf(stderr, "APPNUM RETURN: %s\n", PMIx_Error_string(rc));
    if (PMIX_SUCCESS == rc) {
    	fprintf(stderr, "\t%s\n", PMIx_Value_string(val));
    }

    rc = PMIx_Get(&proc, PMIX_LOCALLDR, NULL, 0, &val);
    fprintf(stderr, "LOCALLDR RETURN: %s\n", PMIx_Error_string(rc));
    if (PMIX_SUCCESS == rc) {
        fprintf(stderr, "\t%s\n", PMIx_Value_string(val));
    }

done:
    PMIx_Finalize(NULL, 0);
    return rc;
}

File: examples/python/toolqry.py
#!/usr/bin/env python3
#
# Copyright (c) 2026      Nanook Consulting  All rights reserved.
# $COPYRIGHT$
#
# Additional copyrights may follow
#
# $HEADER$
#
# Python port of examples/toolqry.c
#
# Attach as a tool, ask the server which namespaces are active, pick one
# that looks like an application job (its name ends in "@N" for non-zero
# N - "@0" is the DVM itself), and then read a couple of job-level values
# out of it.
#
# Pass --system-server to connect to the system-level server.
#
# The namespace list comes back as a data array of infos, each of which is
# itself a data array - so on the Python side it is a list of dicts whose
# values are lists of dicts.

import os
import socket
import sys

from examples import *


def querycbfunc(status, results, mq):
    mq.info = results
    # release the block
    mq.wakeup(status)


def main():
    system_server = False

    # check for directives
    for arg in sys.argv[1:]:
        if "--system-server" == arg:
            system_server = True

    tool = PMIxTool()

    # assign our own name
    hostname = socket.gethostname()
    toolname = os.path.basename(sys.argv[0])
    kptr = "%s.%s.%lu" % (toolname, hostname, os.getpid())
    info = [{'key': PMIX_TOOL_NSPACE, 'value': kptr,
             'val_type': PMIX_STRING},
            {'key': PMIX_TOOL_RANK, 'value': 0,
             'val_type': PMIX_PROC_RANK}]
    # if they want us to use a system server, then say so
    if system_server:
        info.append({'key': PMIX_CONNECT_TO_SYSTEM, 'value': True,
                     'val_type': PMIX_BOOL})

    # init as a tool
    rc, myproc = tool.init(info)
    if PMIX_SUCCESS != rc:
        eprint("PMIx_tool_init failed:", tool.error_string(rc))
        sys.exit(rc)
    proc = {'nspace': "", 'rank': 0}

    # query the list of active nspaces
    mydata = MyQuery()
    query = [{'keys': [PMIX_QUERY_NAMESPACE_INFO], 'qualifiers': None}]
    rc = tool.query_nb(query, querycbfunc, mydata)
    if PMIX_SUCCESS != rc:
        eprint("Client ns %s rank %d: PMIx_Query_info failed: %d"
               % (myproc['nspace'], myproc['rank'], rc))
        return done(tool, rc)
    mydata.wait()
    rc = mydata.status
    # find the response
    if PMIX_SUCCESS != rc:
        eprint("Query returned error:", tool.error_string(mydata.status))
        return done(tool, rc)

    # should be in the first key
    if (not mydata.info
            or PMIX_QUERY_NAMESPACE_INFO != mydata.info[0]['key']):
        eprint("Query returned wrong info key at first posn: %s"
               % (mydata.info[0]['key'] if mydata.info else None))
        return done(tool, PMIX_ERROR)

    darray = mydata.info[0]['value']
    if darray is None or 0 == len(darray.get('array', [])):
        eprint("No namespaces active")
        return done(tool, PMIX_ERROR)

    # find our namespace
    for entry in darray['array']:
        dptr = entry['value']
        if dptr is None or 0 == len(dptr.get('array', [])):
            eprint("Error")
            return done(tool, PMIX_ERROR)
        iptr = dptr['array']
        # looking for nspaces with an "@N" at the end
        nsname = iptr[0]['value']
        _, _, suffix = nsname.rpartition('@')
        if not suffix or suffix == nsname:
            continue
        # ignore the "0" namespace as that is the DVM
        try:
            if 0 == int(suffix):
                continue
        except ValueError:
            continue
        # take this one
        proc = {'nspace': nsname, 'rank': 0}
        eprint("Received nspace %s" % proc['nspace'])
        break

    rc, val = tool.get(proc, PMIX_APPNUM, None)
    eprint("APPNUM RETURN:", tool.error_string(rc))
    if PMIX_SUCCESS == rc:
        _, tmp = tool.value_string(val)
        eprint("\t%s" % tmp)

    rc, val = tool.get(proc, PMIX_LOCALLDR, None)
    eprint("LOCALLDR RETURN:", tool.error_string(rc))
    if PMIX_SUCCESS == rc:
        _, tmp = tool.value_string(val)
        eprint("\t%s" % tmp)

    return done(tool, rc)


def done(tool, rc):
    tool.finalize()
    return rc


if __name__ == '__main__':
    sys.exit(main())
